### PR TITLE
TIP-706: completeness filter

### PR DESCRIPTION
**Pull request by jjanvier**
This PR implements the completeness filter in _ElasticSearch_.
It's behavior is ISO with what we had before, except I added some operators to make the filter more explicit.

Example: completeness EQUALS 100% on the channel tablet 
* in master that means, there is at least one locale of the channel tablet where the completeness EQUALS 100%. Personally, that's not what I expected. I expected all the locales of the channel to be EQUALS to 100%.
* to make it more explicit, I added the operator EQUALS_ON_AT_LEAST_ONE_LOCALE

I did the same for the operators:
* GREATER_THAN
* GREATER_OR_EQUAL_THAN
* LOWER_THAN
* LOWER_OR_EQUAL_THAN
* EQUALS
* NOT_EQUAL

I also moved the _DummyFilter_ in the Enrich bundle as it's used only for UI concerns. Ideally, we should not use that kind of filter at all, but it's not the scope of this PR.  

**The index configuration of the mapping is not done here. It will be done with the number filter, as both filters share the same problematic. We want to merge this PR as fast as possible, as it makes the PIM installable.**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Added integration tests           | Y
| Changelog updated                 | Y
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        $this->supportedFields = $supportedFields;
+        $this->supportedOperators = $supportedOperators;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFieldFilter($field, $operator, $value, $locale = null, $channel = null, $options = [])
+    {
+        $this->checkChannelAndValue($field, $channel, $value);
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php]: You should test if the SearchQueryBuilder is initialized

**Review by BitOne — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                Operators::LOWER_THAN_ON_ALL_LOCALES,
+            ]
+        )) {
+            $this->checkOptions($field, $options);
+            $localeCodes = $options['locales'];
+        } else {
+            $localeCodes = (null !== $locale) ? [$locale] : $this->getChannelByCode($channel)->getLocaleCodes();
+        }
+
+        if (in_array($operator, [Operators::NOT_EQUAL, Operators::NOT_EQUALS_ON_AT_LEAST_ONE_LOCALE])) {
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php]: We should definitely write some comment about this one. Explain what we want and the "trick" with the multiple filterClauses

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+            '!=',
+            '=',
+            '>',
+            '>=',
+            '<=',
+            '<'
+        ];
+
+        $this->beConstructedWith($channelRepository, ['completeness'], $operators);
+        $this->setQueryBuilder($sqb);
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php:43]: You miss the test: It should throw if the query builder is not initialized.

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst [hunk 4]
```diff
@@ -780,11 +780,10 @@ Data model
 
     completenesses:
         print:
-          en_US-number: 100
-          fr_FR-number: 89
+          en_US: 100
+          fr_FR: 89
       ecommerce:
```
Comment (anaelChardan) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst]: Not align with print :)

**Review by anaelChardan — APPROVED** [review]
LGTM

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php (hunk 1 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php]: done

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php (hunk 2 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CompletenessFilter.php]: done

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+
+    function it_throws_an_exception_when_there_is_no_channel()
+    {
+        $this->shouldThrow(InvalidPropertyException::class)->during(
+            'addFieldFilter',
+            ['completeness', Operators::LOWER_THAN_ON_ALL_LOCALES, 100, null, null, ['locales' => ['fr_FR']]]
+        );
+    }
+
+    function it_throws_an_exception_when_there_the_channel_is_unknown_and_when_there_is_no_locale($channelRepository)
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php]: when_there_the => when_the

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+                Operators::LOWER_THAN_ON_AT_LEAST_ONE_LOCALE,
+                100,
+                null,
+                'UNKNOWN',
+                ['locales' => ['fr_FR']]
+            ]
+        );
+    }
+
+    function it_throws_an_exception_when_there_the_data_is_wrong($channelRepository, ChannelInterface $ecommerce)
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php]: same

File: src/Pim/Bundle/EnrichBundle/ProductQueryBuilder/DummyFilter.php [hunk 7]
```diff
@@ -1,6 +1,6 @@
 <?php
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/ProductQueryBuilder/DummyFilter.php]: Shouldn't it be in the Filter namespace ?

File: src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+parameters:
+    pim_enrich.product_query_builder.dummy.class: Pim\Bundle\EnrichBundle\ProductQueryBuilder\DummyFilter
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml]: (if you move DummyFilter to a filter folder update it here)

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/CompletenessesNormalizer.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Normalize the completenesses to the indexing format.
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class CompletenessesNormalizer implements NormalizerInterface
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Normalizer/Indexing/Product/CompletenessesNormalizer.php]: (not in your PR just an idea I have to introduce the word: "CompletenessCollection" to be consistent with the ProductValueCollection)

File: src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/PropertiesNormalizerSpec.php [hunk 10]
```diff
@@ -42,13 +43,48 @@ function it_normalizes_product_properties_with_empty_fields_and_values(
...
+                'family'     => null,
+                'enabled'    => false,
+                'categories' => [],
+                'groups'     => [],
+                'completeness' => [],
+                'values'     => [],
+            ]
+        );
+    }
+    function it_normalizes_product_with_completenesses(
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/PropertiesNormalizerSpec.php]: Add whitespace

**Review by SamirBoulil — APPROVED** [review]
A comments and GTM 👍

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst (hunk 4 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/CompletenessFilterIntegration.php [hunk 11]
```diff
@@ -328,7 +402,7 @@ public function testErrorLocalesIsMalformed()
 
     /**
      * @expectedException \Akeneo\Component\StorageUtils\Exception\InvalidPropertyException
-     * @expectedExceptionMessage Property "completeness" expects a valid scope.
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/CompletenessFilterIntegration.php]: nope, we will have problem on API if you change scope to channel.
PO wanted to have "scope" everywhere to be consistent

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php (hunk 5 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php]: done

File: src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php (hunk 6 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/spec/Elasticsearch/Filter/CompletenessFilterSpec.php]: done

File: src/Pim/Bundle/EnrichBundle/ProductQueryBuilder/DummyFilter.php (hunk 7 above)
Comment (jjanvier) [reply src/Pim/Bundle/EnrichBundle/ProductQueryBuilder/DummyFilter.php]: ok, if you wnat

File: src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml (hunk 8 above)
Comment (jjanvier) [reply src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml]: done

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/CompletenessesNormalizer.php (hunk 9 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Normalizer/Indexing/Product/CompletenessesNormalizer.php]: done

File: src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/PropertiesNormalizerSpec.php (hunk 10 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/PropertiesNormalizerSpec.php]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/CompletenessFilterIntegration.php (hunk 11 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/CompletenessFilterIntegration.php]: reverted

**Pull request merged**